iothub_client/samples/pnp/common/pnp_protocol_helpers.c

```diff
@@ -0,0 +1,327 @@
+        jsonToSend = STRING_construct_sprintf(g_propertyWithResponseSchemaWithoutComponent, propertyName, propertyValue, result, description, ackVersion);
+    }
+    else 
+    {
+       jsonToSend = STRING_construct_sprintf(g_propertyWithResponseSchemaWithComponent, componentName, propertyName, propertyValue, result, description, ackVersion);
+    }
+
+    if (jsonToSend == NULL)
+    {
+        LogError("Unable to allocate JSON buffer");
+    }
+
+    return jsonToSend;    
+}
+
+void PnPHelper_ParseCommandName(const char* deviceMethodName, unsigned const char** componentName, size_t* componentNameSize, const char** pnpCommandName)
+{
+    const char* separator;
+
+    if ((separator = strchr(deviceMethodName, g_commandSeparator)) != NULL)
+    {
+        // If a separator character is present in the device method name, then a command on a subcomponent of 
+        // the model is being targeted (e.g. thermostat1*getMaxMinReport).
+        *componentName = (unsigned const char*)deviceMethodName;
+        *componentNameSize = separator - deviceMethodName;
+        *pnpCommandName = separator + 1;
+    }
+    else 
+    {
+        // The separator character is optional.  If it is not present, it indicates a command of the root 
+        // component and not a subcomponent (e.g. "reboot").
+        *componentName = NULL;
+        *componentNameSize = 0;
+        *pnpCommandName = deviceMethodName;
+    }
+}
+
+IOTHUB_MESSAGE_HANDLE PnPHelper_CreateTelemetryMessageHandle(const char* componentName, const char* telemetryData) 
+{
+    IOTHUB_MESSAGE_HANDLE messageHandle;
+    IOTHUB_MESSAGE_RESULT iothubMessageResult;
+    bool result;
+    
+    if ((messageHandle = IoTHubMessage_CreateFromString(telemetryData)) == NULL)
+    {
+        LogError("IoTHubMessage_CreateFromString failed");
+        result = false;
+    }
+    // If the component will be used, then specify this as a property of the message.
+    else if ((componentName != NULL) && (iothubMessageResult = IoTHubMessage_SetProperty(messageHandle, PnP_TelemetryComponentProperty, componentName)) != IOTHUB_MESSAGE_OK)
+    {
+        LogError("IoTHubMessage_SetProperty=%s failed, error=%d", PnP_TelemetryComponentProperty, iothubMessageResult);
+        result = false;
+    }
+    else
+    {
+        result = true;
+    }
+
+    if ((result == false) && (messageHandle != NULL))
+    {
+        IoTHubMessage_Destroy(messageHandle);
+        messageHandle = NULL;
+    }
+    
+    return messageHandle;
+}
+
+//
+// VisitComponentProperties visits each sub element of the the given objectName in the desired JSON.  Each of these sub elements corresponds to
+// a property of this component, which we'll invoke the application's pnpPropertyCallback to inform.
+// 
+static void VisitComponentProperties(const char* objectName, JSON_Value* value, int version, PnPHelperPropertyCallbackFunction pnpPropertyCallback, void* userContextCallback)
+{
+    JSON_Object* object = json_value_get_object(value);
```

Comment:
validate object?

Reply:
I'd considered but I don't technically have to here.  If `object` is NULL, then the subsequent call to `json_object_get_count` will end up returning a 0 and the following loop will be a giant no-op.  And note in practice this call shouldn't ever fail anyway, as we're already marked as a static so we can trust caller in our function and it only calls if it knows we're an object.